src/mqt/problemsolver/partialcompiler/qaoa.py

```diff
@@ -0,0 +1,145 @@
+        considered_following_qubits: int = 3,
+    ):
+        self.num_qubits = num_qubits
+        self.repetitions = repetitions
+
+        assert 0 <= remove_probability <= 1
+        self.sample_probability = remove_probability
+        np.random.seed(42)
+
+        self.backend = get_backend(num_qubits)
+
+        qc, qc_baseline, remove_gates = self.get_uncompiled_circuits(considered_following_qubits)
+        self.qc = qc  # QC with all gates
+        self.qc_baseline = qc_baseline  # QC with only the sampled gates
+        self.remove_gates = remove_gates  # List of booleans indicating whether a gate should be removed
+        self.qc_compiled = self.compile_qc(baseline=False, opt_level=3)  # Compiled QC with all gates
+        self.to_be_removed_gates_indices = self.get_to_be_removed_gate_indices()  # Indices of the gates to be checked
+
+    def get_uncompiled_circuits(
+        self, considered_following_qubits: int
+    ) -> tuple[QuantumCircuit, QuantumCircuit, list[bool | str]]:
+        """Returns the uncompiled circuits (both with only the actual needed two-qubit gates and with all possible
+        two-qubit gates) and the list of gates to be removed."""
+
+        qc = QuantumCircuit(self.num_qubits)  # QC with all gates
+        qc_baseline = QuantumCircuit(self.num_qubits)  # QC with only the sampled gates
+        qc.h(range(self.num_qubits))
+        qc_baseline.h(range(self.num_qubits))
+
+        remove_gates: list[bool | str] = []
+        parameter_counter = 0
+        tmp_len = -1
+
+        # Iterate over all QAOA layers
+        for k in range(self.repetitions):
+            if k == 1:
+                tmp_len = len(remove_gates)  # Number of parameterized gates in the first layer
+
+            # Iterate over all possible two-qubit gates
+            for i in range(self.num_qubits):
+                # considered_following_qubits describes the number of to be considered following qubits
+                for j in range(i + 1, i + 1 + considered_following_qubits):
+                    if j >= self.num_qubits:
+                        break
+                    p = Parameter(f"a_{parameter_counter}")
+                    qc.rzz(p, i, j)
+                    # Sample whether the gate should be removed for the first layer
+                    if k == 0:
+                        if np.random.random() < self.sample_probability:
+                            remove_gates.append(p.name)
+                        else:
+                            remove_gates.append(False)
+                            qc_baseline.rzz(p, i, j)
+                    # For all other layers, check whether the gate should be removed
+                    elif remove_gates[parameter_counter - k * tmp_len]:
+                        remove_gates.append(p.name)
+                    else:
+                        remove_gates.append(False)
+                        qc_baseline.rzz(p, i, j)
+                    parameter_counter += 1
```

Comment:
Wouldn't it require way less special handling to construct one QAOA layer and then just replicate that?

Reply:
The problem with that would be, that the RZ-Gate parameters would be the exact same ones for every iteration. While it obviously would not matter for our compilation scheme, it feels wrong to me to create a QAOA circuit which could not be used as intended.

Reply:
I would have guessed there is some way to use separate parameters for each iteration. Anyway, it's not really going to affect the performance.

Reply:
I played around with it and ended up with a method which gets the starting index, the repetition number, and the `remove_gate` list to create a new QAOA iteration while I didn't get rid of the case differentiation (first repetition determines the to be removed gates, all others follow the same pattern). 

Therefore, I suggest to leave it as is for now.